src/profile/context/ProfileContext.tsx

```diff
@@ -0,0 +1,173 @@
+    },
+  });
+  const [profileData, updateData] = useState(data);
+  const loadTracker = useRef({
+    started: false,
+    complete: false,
+    loading: false,
+  });
+
+  const updateProfileData: ProfileContextData['updateProfileData'] = useCallback(
+    newProfileData => {
+      loadTracker.current.complete = true;
+      updateData(newProfileData);
+    },
+    [updateData]
+  );
+
+  const fetchDataIfNotLoaded: ProfileContextData['fetch'] = useCallback(() => {
+    if (loadTracker.current.started) {
+      return;
+    }
+    loadTracker.current.started = true;
+    loadTracker.current.loading = true;
+    fetch();
+  }, [fetch, loadTracker]);
+
+  const refetchDataIfPossible: ProfileContextData['refetch'] = useCallback(() => {
+    if (!refetch) {
+      return Promise.reject(
+        new Error('Cannot refetch before first fetch is done. Fetch first.')
+      );
+    }
+    loadTracker.current.loading = true;
+    return refetch();
+  }, [refetch]);
+
+  if (loadTracker.current.loading && loading === false && data) {
+    loadTracker.current.loading = false;
+    loadTracker.current.complete = true;
+    updateProfileData(data);
+  } else if (!profileData && data && !loading) {
+    loadTracker.current.started = true;
+    loadTracker.current.complete = true;
+    loadTracker.current.loading = false;
+    updateProfileData(data);
+  }
+
+  const contextData: ProfileContextData = {
+    data: profileData,
+    updateProfileData,
+    loading,
+    error,
+    addErrorListener: f => {
+      errorListeners.current.listeners.add(f);
+      return () => errorListeners.current.listeners.delete(f);
+    },
+    fetch: fetchDataIfNotLoaded,
+    refetch: refetchDataIfPossible,
+    isInitialized: loadTracker.current.started,
+    isComplete: loadTracker.current.complete,
+    getName: preferNickOrGivenName => {
+      if (!profileData || !profileData.myProfile) {
+        return '';
+      }
+      const verifiedPersonalInformation = getVerifiedPersonalInformation(
+        profileData
+      );
+      if (verifiedPersonalInformation) {
+        return `${
+          preferNickOrGivenName && verifiedPersonalInformation.givenName
+            ? verifiedPersonalInformation.givenName
+            : verifiedPersonalInformation.firstName
+        } ${verifiedPersonalInformation.lastName}`;
+      } else {
+        const source = profileData.myProfile;
+        return preferNickOrGivenName && source.nickname
+          ? source.nickname
+          : `${source.firstName} ${source.lastName}`;
+      }
+    },
+  };
+
+  return (
+    <ProfileContext.Provider value={contextData}>
+      {children}
+    </ProfileContext.Provider>
+  );
+};
+
+export const useProfileListener = (listener: ErrorListener): void => {
+  const { addErrorListener } = useContext(ProfileContext);
+
+  const disposeListener = useMemo(
+    () => addErrorListener(listener),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    []
+  );
```

Comment:
Dependencies are not included in the array, because it is important that listener is added exactly once and useMemo should never be run again. Empty dependency array takes care of that.

But "addListener" is now added as dependency, because I changed the code in ProfileContext and it never changes.

"listener" is now also a dependency, because it is passed as a never changing function created with new hook: usePersistFn. It always stores the last version of the listener and makes sure it is called. This way re-render loops or multiple identical listeners cant happen.

Reply:
Good that the dependencies are now there.

But isn't it now (and used to be too) so that the user of the `useProfileErrorListener` hook can't change the listener even if they wanted to? Is it this hook's job to decide that the user can not change the listener afterwards? Shouldn't it be the caller's freedom to decide if they want to change listeners at some point?

I'm perhaps pondering general hook design "guidelines" here, learning how they should generally be implemented. Within a single project like this anything is off course possible, as long as the consequences are understood.

Reply:
Listener changes every time. The function passed to useProfileErrorListener() is always stored over the previous one in usePersistFn(). 

https://github.com/alibaba/hooks/blob/master/packages/hooks/src/usePersistFn/index.ts

But changing it does not trigger the useMemo and useEffect.

Alternative to usePersistFn would be to wrap the listener with "useCallback" where ever useProfileErrorListener() is called from. But usePersistFn simplifies the process and "useCallback" is not required every time and listener can be any kind of function.